### PR TITLE
Quantized Interpolate Kernel(upsample_bilinear2d)

### PR DESCRIPTION
**Pull request by lly-zero-one**
We implement the quantized upsample_bilinear2d case for interpolate kernel in this PR.

For nhwc performance improvement:
import torch, time

for dtype in [torch.qint8, torch.quint8, torch.qint32]:
    print('****', str(dtype), '*****')
    x = torch.rand(1, 56, 56, 256)

    q_x = torch.quantize_per_tensor(x, 0.5, 1, dtype)
    q_x = q_x.permute([0, 3, 1, 2])

    x = x.permute([0, 3, 1, 2])

    NITER = 100

    s = time.time()
    for i in range(NITER):
        float_out = torch.nn.functional.interpolate(x, size=5, scale_factor=None, mode="bilinear", align_corners=True)
    time_per_iter_float = (time.time() - s) / NITER

    s = time.time()
    for i in range(NITER):
        quant_out = torch.nn.quantized.functional.interpolate(q_x, size=5, scale_factor=None, mode="bilinear", align_corners=True)
    time_per_iter_quant = (time.time() - s) / NITER

    ref_quantized = torch.quantize_per_tensor(float_out, 0.5, 1, dtype)
    #  torch.testing.assert_allclose(ref_quantized.dequantize(), quant_out.dequantize())

    print('time/iter ms (float)', 'time/iter ms (quant)', 'quant/float', sep='\t')
    print(time_per_iter_float * 1000, time_per_iter_quant * 1000, time_per_iter_quant / time_per_iter_float, sep='\t')

    bytes_float = (x.numel() + float_out.numel()) * x.element_size()
    bytes_quant = (q_x.numel() + quant_out.numel()) * q_x.element_size()

    float_bw_gbps = bytes_float / time_per_iter_float / 1e9
    quant_bw_gbps = bytes_quant / time_per_iter_quant / 1e9

    print('GB/s float', 'GB/s quant', sep='\t')
    print(float_bw_gbps, quant_bw_gbps, sep='\t')

===========without nhwc handling===========
**** torch.qint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
1.999044418334961       2.5860953330993652      1.2936657681940702
GB/s float      GB/s quant
1.6192056416115257      0.3129103516188541
**** torch.quint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.02730655670166        2.6061582565307617      1.2855274639721328
GB/s float      GB/s quant
1.596632728927902       0.3105014816242217
**** torch.qint32 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.0180463790893555      2.4047350883483887      1.1916153728010588
GB/s float      GB/s quant
1.603959172365819       1.3460376636426636

===========with nhwc handling===========

**** torch.qint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.0913314819335938      0.09696483612060547     0.04636512047863123
GB/s float      GB/s quant
1.5477527249803915      8.345458337015
**** torch.quint8 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.1065664291381836      0.09959936141967773     0.04728042754408879
GB/s float      GB/s quant
1.5365591871338384      8.124710725706763
**** torch.qint32 *****
time/iter ms (float)    time/iter ms (quant)    quant/float
2.044203281402588       0.6003522872924805      0.29368521846837126
GB/s float      GB/s quant
1.5834354779917448      5.391607675216635

### CODE REVIEW

**Comment by llyfacebook** [comment]
TODO: for nhwc layout, we could use FMA to get some parallelism.

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 1]
```diff
@@ -447,6 +449,155 @@ void qavg_pool2d_nhwc_kernel(
...
+          h1lambda_v * (w0lambda_v * pos1_fp_v[2] + w1lambda_v * pos1_fp_v[3]);
+      float result_fp[vec_width];
+      result.store(result_fp);
+      vec256::QuantizeAvx2<T>(
+          result_fp, pos2, vec_width, multiplier, output_zero_point);
+      pos1 += vec_width;
+      pos2 += vec_width;
+    }
+  }
+#endif
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Make an `#else` branch here with a `TORCH_INTERNAL_ASSERT(false` to catch bugs with the preprocessor conditionals in the future

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 2]
```diff
@@ -447,6 +449,155 @@ void qavg_pool2d_nhwc_kernel(
   });
 }
 
+template <typename T>
+void do_bilinear_on_AVX2(
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Add something to this function name to indicate it's quantized

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 3]
```diff
@@ -447,6 +449,155 @@ void qavg_pool2d_nhwc_kernel(
...
+      vec256::QuantizeAvx2<T>(
+          result_fp, pos2, vec_width, multiplier, output_zero_point);
+      pos1 += vec_width;
+      pos2 += vec_width;
+    }
+  }
+#endif
+}
+
+void upsample_bilinear2d_nhwc_kernel(
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Same here, indicate that this function is quantized

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 4]
```diff
@@ -447,6 +449,155 @@ void qavg_pool2d_nhwc_kernel(
...
+              const int64_t w1 = w1r;
+              const int64_t w1p = (w1 < input_width - 1) ? 1 : 0;
+
+              const float w1lambda = w1r - w1;
+              const float w0lambda = static_cast<float>(1.) - w1lambda;
+
+              int64_t c = 0;
+              // We use float32 to do the computation
+              const auto* pos1 = i_p + (h1 * input_width + w1) * channels;
+              auto* pos2 = o_p + (h2 * output_width + w2) * channels;
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Don't use `auto` in these two lines

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 5]
```diff
@@ -447,6 +449,155 @@ void qavg_pool2d_nhwc_kernel(
...
+                  channels,
+                  output.q_zero_point(),
+                  multiplier,
+                  h0lambda,
+                  h1lambda,
+                  w0lambda,
+                  w1lambda,
+                  h1p,
+                  w1p);
+              for (; c < channels; ++c) {
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: I don't like that `c` is taken and updated by reference in `do_bilinear_on_AVX2`, it's not readily apparent `c` is updated by reading the code here. Can you instead return the new value of `c` from `do_bilinear_on_AVX2`?

File: aten/src/ATen/native/quantized/cpu/qupsample_bilinear2d.cpp [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    for (int64_t h2 = 0; h2 < output_height; ++h2) {
+      const int64_t h1 = h2;
+
+      for (int64_t w2 = 0; w2 < output_width; ++w2) {
+        const int64_t w1 = w2;
+        const auto* pos1 = &i_p[h1 * input_width + w1];
+        auto* pos2 = &o_p[h2 * output_width + w2];
+
+        for (int64_t c = 0; c < channels; ++c) {
+          pos2[0] = pos1[0];
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/qupsample_bilinear2d.cpp]: Same comment as the other PR: this section is autovectorizable and should probably go in kernels/

File: aten/src/ATen/native/quantized/cpu/qupsample_bilinear2d.cpp [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+      const auto* pos1 = i_p + h1 * input_width + w1;
+      auto* pos2 = o_p + h2 * output_width + w2;
+
+      for (int64_t c = 0; c < channels; ++c) {
+        float result = h0lambda * (w0lambda * pos1[0] + w1lambda * pos1[w1p]) +
+            h1lambda *
+                (w0lambda * pos1[h1p * input_width] +
+                 w1lambda * pos1[h1p * input_width + w1p]);
+        // requantization
+        pos2[0] = static_cast<typename scalar_t::underlying>(std::min<int32_t>(
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/qupsample_bilinear2d.cpp]: `at::quantize_val`?

**Review by jamesr66a — COMMENTED** [review]
Looks OK, but I'm concerned about all the branches with no `else` statement. If these are precondition checks, they should be `TORCH_INTERNAL_ASSERT`s rather than branches. If they are not, you need an `else`

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 1 above)
Comment (llyfacebook) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: I think it is not supposed to fail in the else branch. We just ignore this code for non-AVX2 or windows build.

File: aten/src/ATen/native/quantized/cpu/qupsample_bilinear2d.cpp (hunk 6 above)
Comment (llyfacebook) [reply aten/src/ATen/native/quantized/cpu/qupsample_bilinear2d.cpp]: I think memcpy makes sense. But it is already a library call, which should always have its own dispatching for AVX/AVX2. So I still think we should keep this part here.

Comment (jamesr66a) [reply aten/src/ATen/native/quantized/cpu/qupsample_bilinear2d.cpp]: `memcpy` can be replaced with a builtin and inlined by the compiler. We should not be handrolling memcpy like this.

**Review by dzhulgakov — COMMENTED** [review]
You need to provide implementation for non-AVX2 too or at least generate a reasonable error message (which we do with fbgemm ops)

**Comment by llyfacebook** [comment]
> You need to provide implementation for non-AVX2 too or at least generate a reasonable error message (which we do with fbgemm ops)

I think it is already there. The loop after the function call is the default implementation.

**Review by jamesr66a — APPROVED** [review]
I think @dzhulgakov's comment is still valid. It's very hard (still) to figure out that the implementation is of the form

```
int64_t c = 0;
#ifdef __AVX2__
for (; c + vec_width < channels; c+= vec_width) {
  ...
}
#endif
for (; c < channels; ++c) {
...
}
```

Which I think is very clear to people that the scalar implementation is the fallback. @llyfacebook can you add some explicit comments stating why we've factored this out as a function and explaining what it would look like in the ideal case?

Approved otherwise

**Pull request closed without merging**